src/main/kotlin/org/janelia/saalfeldlab/labels/blocks/n5/LabelBlockLookupFromN5.kt

```diff
@@ -55,103 +55,125 @@ class LabelBlockLookupFromN5(
 			return map
 		}
 
-		private fun toBytes(map: Map<Long, Array<Interval>>): ByteArray {
-			val sizeInBytes = map.values.stream().mapToInt { java.lang.Long.BYTES + Integer.BYTES + SINGLE_ENTRY_BYTE_SIZE * it.size }.sum()
+		private fun toBytes(map: TLongObjectHashMap<Array<Interval>>): ByteArray {
+			val sizeInBytes = map.valueCollection().stream().mapToInt { java.lang.Long.BYTES + Integer.BYTES + SINGLE_ENTRY_BYTE_SIZE * it.size }.sum()
 			val bytes = ByteArray(sizeInBytes)
 			val bb = ByteBuffer.wrap(bytes)
-			for (entry in map) {
-				bb.putLong(entry.key)
-				bb.putInt(entry.value.size)
-				for (interval in entry.value) {
+			map.forEachEntry { key, value ->
+				bb.putLong(key)
+				bb.putInt(value.size)
+				for (interval in value) {
 					bb.putLong(interval.min(0))
 					bb.putLong(interval.min(1))
 					bb.putLong(interval.min(2))
 					bb.putLong(interval.max(0))
 					bb.putLong(interval.max(1))
 					bb.putLong(interval.max(2))
 				}
+				true
 			}
 			return bytes
 		}
 	}
 
 	private val attributes = mutableMapOf<Int, DatasetAttributes>()
 
+	private val cache = TIntObjectHashMap<TLongObjectHashMap<TLongObjectHashMap<Array<Interval>>>>() // scale -> block index -> id -> intervals
 
 	@Throws(IOException::class)
-	fun set(level: Int, ids: Map<Long, Array<Interval>>) {
-
-
-		val dataset = String.format(scaleDatasetPattern, level)
-		val attributes = this.attributes.getOrPut(level, { n5().getDatasetAttributes(dataset) })
-		val stepSize = attributes.blockSize[0]
-
-		val mapByBlock = mutableMapOf<Long, MutableMap<Long, Array<Interval>>>()
-
-		for (entry in ids)
-			mapByBlock.computeIfAbsent((entry.key / stepSize) * stepSize, { mutableMapOf() })[entry.key] = entry.value
-
+	@Synchronized
+	fun set(level: Int, ids: TLongObjectHashMap<Array<Interval>>) {
+		val mapByBlock = mutableMapOf<Long, TLongObjectHashMap<Array<Interval>>>()
+		ids.forEachEntry { key, value ->
+			mapByBlock.computeIfAbsent(getBlockId(level, key)) { TLongObjectHashMap() }.put(key, value)
+			true
+		}
 		for (m in mapByBlock)
-			writeMap(level, m.key, m.value)
-
+			writeAndCacheMap(level, m.key, m.value)
 	}
 
-
 	@Throws(IOException::class)
+	@Synchronized
 	override fun read(level: Int, id: Long): Array<Interval> {
-		LOG.debug("Reading id {} for level={}", id, level);
-		val map = readMap(level, id) ?: mutableMapOf()
-		return map.getOrElse(id, { emptyArray() })
+		LOG.debug("Reading id {} for level={}", id, level)
+		val map = getMap(level, getBlockId(level, id))
+		return map.get(id)
+	}
 
+	@Throws(IOException::class)
+	@Synchronized
+	override fun write(level: Int, id: Long, vararg intervals: Interval) {
+		val blockId = getBlockId(level, id)
+		val map = readAndCacheMap(level, blockId)
+		map.put(id, arrayOf(*intervals))
+		writeAndCacheMap(level, blockId, map)
```

Comment:
If I understand correctly: If we use imglib2 cache, we could just write and then [`Invalidate.invalidateIf`](https://github.com/imglib/imglib2-cache/blob/fe7ee81cda06a0c6bf48ae3edbd2f3c2f1067460/src/main/java/net/imglib2/cache/Invalidate.java#L70) the matching entries in the cache.